Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs

```diff
@@ -21,24 +21,38 @@ namespace Neo4j.Driver.Internal.Result;
 
 internal class Record : IRecord
 {
-    public Record(string[] keys, object[] values)
-    {
-        Throw.ProtocolException.IfNotEqual(keys.Length, values.Length, nameof(keys), nameof(values));
-
-        var valueKeys = new Dictionary<string, object>();
+    private readonly Dictionary<string, int> _fieldIndexes;
+    private readonly object[] _values;
 
-        for (var i = 0; i < keys.Length; i++)
+    public Record(string[] keys, Dictionary<string, int> fieldIndexes, object[] values)
+    {
+        if (keys.Length != values.Length)
         {
-            valueKeys.Add(keys[i], values[i]);
+            throw new ProtocolException(
+                $"{nameof(keys)} ({keys.Length.ToString()}) does not equal to {nameof(values)} ({values.Length.ToString()})");
         }
-
-        Values = valueKeys;
+        
+        _fieldIndexes = fieldIndexes;
+        _values = values;
         Keys = keys;
     }
 
-    public object this[int index] => Values[Keys[index]];
-    public object this[string key] => Values[key];
+    public object this[int index] => _values[index];
+    public object this[string key] => _values[_fieldIndexes[key]];
+
+    public IReadOnlyDictionary<string, object> Values
+    {
+        get
+        {
+            var valueKeys = new Dictionary<string, object>(_values.Length);
```

Comment:
we are building this once per call, it could be stored for reuse for the user code.
```csharp
var a = record.Values["a"];
var b = record.Values["b"];
```